source/extensions/filters/http/common/jwks_fetcher.cc

```diff
@@ -0,0 +1,91 @@
+#include "extensions/filters/http/common/jwks_fetcher.h"
+
+#include "common/common/enum_to_int.h"
+#include "common/http/headers.h"
+#include "common/http/utility.h"
+
+#include "jwt_verify_lib/status.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Common {
+namespace {
+class JwksFetcherImpl : public JwksFetcher,
+                        public Logger::Loggable<Logger::Id::filter>,
+                        public Http::AsyncClient::Callbacks {
+private:
+  Upstream::ClusterManager& cm_;
+  JwksFetcher::JwksReceiver* receiver_ = nullptr;
+  const ::envoy::api::v2::core::HttpUri* uri_ = nullptr;
+  Http::AsyncClient::Request* request_ = nullptr;
+
+public:
+  JwksFetcherImpl(Upstream::ClusterManager& cm) : cm_(cm) { ENVOY_LOG(trace, "{}", __func__); }
+
+  void cancel() {
+    if (request_) {
+      request_->cancel();
+      request_ = nullptr;
+      ENVOY_LOG(debug, "fetch pubkey [uri = {}]: canceled", uri_->uri());
+    }
+  }
+
+  void fetch(const ::envoy::api::v2::core::HttpUri& uri, JwksFetcher::JwksReceiver& receiver) {
+    ENVOY_LOG(trace, "{}", __func__);
+    receiver_ = &receiver;
```

Comment:
What if `receiver_` is already set? What is the contract on multiple fetches?

Reply:
There's a description of the use of the JwksFetcher interface just before the definition in the header file. Would you like further commentary closer to the fetch() method?

Reply:
Can you `ASSERT` here that `recever_ == nullptr`, i.e. we catch in debug builds violations of the contract.